server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/reactivestreams/FileRangePublisher.scala

```diff
@@ -0,0 +1,84 @@
+package sttp.tapir.server.netty.internal.reactivestreams
+
+import io.netty.buffer.Unpooled
+import io.netty.handler.codec.http.{DefaultHttpContent, HttpContent}
+import org.reactivestreams.{Publisher, Subscriber, Subscription}
+import sttp.tapir.FileRange
+
+import java.nio.ByteBuffer
+import java.nio.channels.{AsynchronousFileChannel, CompletionHandler}
+import java.nio.file.StandardOpenOption
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicLong}
+
+class FileRangePublisher(fileRange: FileRange, chunkSize: Int) extends Publisher[HttpContent] {
+  override def subscribe(subscriber: Subscriber[_ >: HttpContent]): Unit = {
+    if (subscriber == null) throw new NullPointerException("Subscriber cannot be null")
+    val subscription = new FileRangeSubscription(subscriber, fileRange, chunkSize)
+    subscriber.onSubscribe(subscription)
+  }
+
+  private class FileRangeSubscription(subscriber: Subscriber[_ >: HttpContent], fileRange: FileRange, chunkSize: Int) extends Subscription {
+    private lazy val channel: AsynchronousFileChannel = AsynchronousFileChannel.open(fileRange.file.toPath(), StandardOpenOption.READ)
+    private val demand = new AtomicLong(0L)
+    private val position = new AtomicLong(fileRange.range.flatMap(_.start).getOrElse(0L))
+    private val buffer: ByteBuffer = ByteBuffer.allocate(chunkSize)
+    private val isCompleted = new AtomicBoolean(false)
+    private val readingInProgress = new AtomicBoolean(false)
+
+    override def request(n: Long): Unit = {
+      if (n <= 0) subscriber.onError(new IllegalArgumentException("§3.9: n must be greater than 0"))
+      else {
+        demand.addAndGet(n)
+        readNextChunkIfNeeded()
+      }
+    }
+
+    private def readNextChunkIfNeeded(): Unit = {
+      if (demand.get() > 0 && !isCompleted.get() && readingInProgress.compareAndSet(false, true)) {
```

Comment:
The main idea here is to not allow more than 1 thread to call `channel.read`. I think this constraint is provided by `readingInProgress.compareAndSet`, so there's either:
a) a thread that comes through from `request(n)` if nothing was in progress 
b) a thread that comes through from previous `read` `completed` callback

Also, reactive streams specs guarantee that `request` is never called before previous `request()` returns.

Reply:
looks good - I'd add the above as a comment - stating an invariant (only one thread can run the logic below), and how it's guarded. Plus the only possibility of a race - when demand & chunk completion happen at the same time